eip-0005.md

```diff
@@ -0,0 +1,170 @@
+ jar library generated by Scala2.12 is not supported (Android, native, web, etc.)
+ - Once generated the contract template becomes self-contained and can be stored,
+ transfered, parsed, executed etc across platforms, tools and applications.
+
+## Contract Template Serialization Format
+
+Each contract template can be represented using an array of bytes. The content of the
+array is the following sequence of data fields. Note, the Field Name is not saved and used
+here only for description of the corresponding content.
+Binary serialization format is suitable for compact storage and wire transfers.
+
+
+Field Name     |  Format            | Example                | Description
+---------------|--------------------|------------------------|-----------
+`NameLength`   | `VLQ(UInt)`        | 18                     | Length of `Name` bytes 
+`Name`         | `Bytes`            | "rewardOutputScript"   | User readable name (string bytes in UTF-8 encoding)
+`DescLength`   | `VLQ(UInt)`        | 20                     | Length of `Description` bytes 
+`Description`  | `Bytes`            | "holds mining rewards" | User readable contract description (string bytes in UTF-8 encoding)
+`ConstLength`  | `VLQ(UInt)`        | 2                      | Number of items in `Constants` 
+`Constants`    | `Const+`           | (*)                    | Constant serialization bytes (see _Constant Serialization_ section of [ErgoTree Spec](https://ergoplatform.org/docs/ErgoTree.pdf)
+`Template`     | `Expr`             | (**)                   | Contract Template bytes with parameter placeholders (see _Expression Serialization_ section of [ErgoTree Spec](https://ergoplatform.org/docs/ErgoTree.pdf)
+`ParamCount`   | `VLQ(UInt)`        | 1                      | Number of named template parameters
+`Parameters`   | `Parameter+`       | delta: Int             | Typed template parameter (see [Parameter Serialization Format](#parameter-serialization-format) section)
+
+_(*)_
+IR: 
+```
+Seq(
+  Constant(720, SInt), 
+  Constant(SigmaProp(ProveDlog(ECPoint(79be66, 483ada,...))), SSigmaProp)
+)
+```
+Hex: "0x0204a00b08cd0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798"
+
+_(**)_
+IR: 
+```
+SigmaAnd(Seq(
+    BoolToSigmaProp(GE(Height,Plus(SelectField(ExtractCreationInfo(Self),1), ConstantPlaceholder(0,SInt)))),
+    ConstantPlaceholder(1,SSigmaProp)))
+)
+```
+Hex: "ea02d192a39a8cc7a70173007301"
+
+## Parameter Serialization Format
+
+Parameters specify additional meta information for the contract template. The number of
+parameters can be less then equal the number of segregated constants. This is because some
+constant values may not be changed without breaking contract logic and hence they cannot
+be used as template parameters.
+
+Field Name     |  Format            | Example             | Description
+---------------|--------------------|---------------------|-----------
```

Comment:
Which fields are mandatory and which are not?

Reply:
described